snowflake/exec.go

```diff
@@ -1,28 +1,73 @@
 package snowflake
 
 import (
+	"context"
 	"database/sql"
 
 	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
-func ExecNoRows(db *sql.DB, query string) (sql.Result, error) {
-	logrus.Debug("[DEBUG] exec stmt ", query)
+func ExecNoRows(ctx context.Context, db *sql.DB, query string, args ...interface{}) (sql.Result, error) {
+	logrus.Debugf("[DEBUG] exec stmt (%s)", query)
 
-	return db.Exec(query)
+	// if no args, then don't prepare:
+	if len(args) == 0 {
+		return db.ExecContext(ctx, query)
+	}
+
+	// Prepare the statement
+	preparedStmt, err := db.PrepareContext(ctx, query)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to prepare sql statement (%s)", query)
+	}
+	defer preparedStmt.Close()
+
+	sqlResult, err := preparedStmt.ExecContext(ctx, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to execute prepared statement")
+	}
+
+	// Run statement with arguments
+	return sqlResult, nil
 }
 
-func ExecMulti(db *sql.DB, queries []string) error {
+func ExecMulti(ctx context.Context, db *sql.DB, queries []string, args ...[]interface{}) error {
 	logrus.Debug("[DEBUG] exec stmts ", queries)
 
-	tx, err := db.Begin()
+	tx, err := db.BeginTx(ctx, nil)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Unable to begin transaction with database")
 	}
 
-	for _, query := range queries {
-		_, err = tx.Exec(query)
+	// NOTE(aku): do we need to store these statements one-by-one so we can programmatically close them all?
+	// preparedStmts := []*sql.Stmt{}
+
+	for i, query := range queries {
+		// Prepare the statement
+		logrus.Debugf("[DEBUG] exec (%s) ", query)
+		// if no args, then don't prepare:
+		if len(args[i]) == 0 {
+			_, err = db.ExecContext(ctx, query)
+			if err != nil {
+				return errors.Wrap(err, "Unable to execute statement")
+			}
+
+			return nil
+		}
+
+		preparedStmt, err := tx.PrepareContext(ctx, query)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to prepare query (%s)", query)
+		}
+		defer preparedStmt.Close()
+		// TODO(aku): do we need to do this?
+		// preparedStmts = append(preparedStmts, preparedStmt)
+		// defer preparedStmts[i].Close()
```

Comment:
I'm not sure how to test this: if I'm in a for loop with a new query each iteration, will they all get closed if I do this? I'm worried because I don't know how to "test" this kind of thing. I tried to write an alternative (commented out)

Reply:
I think what you have is fine. The defer will get called at the end of the function. If the number of queries is super large, it might be better to just close the prepared statement right away. Otherwise, you can just keep what you have.